doc/lspconfig.txt

```diff
@@ -1,242 +1,628 @@
+servers for which there is not a configuration, it is necessary to define a
+configuration directly. This can be useful if there is an outstanding PR that
+is in review, or when developing a language server that is unavailable
+publicly.
 
+The `configs` module is a singleton where configs are defined. The schema for
+validating using `vim.validate` is:
 >
-    lua << EOF    
-    vim.lsp.set_log_level("debug")    
-    EOF
+    configs.SERVER_NAME = {
+      default_config = {'t'};
+      on_new_config = {'f', true};
+      on_attach = {'f', true};
+      commands = {'t', true};
+      docs = {'t', true};
+    }
 <
+Where the structure of the docs table is as follows:
+>
+    docs = {
+      description = {'s', true};
+      default_config = {'t', true};
+    }
+<
+`commands` is a map of `name:definition` key:value pairs, where `definition`
+is a list whose first value is a function implementing the command and the
+rest are either array values which will be formed into flags for the command
+or special keys like `description`. Example:
 
+>
+  commands = {
+    TexlabBuild = {
+      function()
+        buf_build(0)
+      end;
+      "-range";
+      description = "Build the current buffer";
+    };
+  };
+<
 
-Attempt to run the language server, and open the log with:
+The `configs.__newindex` metamethod consumes the config definition and returns
+an object with a `setup()` method, to be invoked by users:
+>
+    require'lspconfig'.SERVER_NAME.setup{}
+
+After you set `configs.SERVER_NAME` you can add arbitrary language-specific
+functions to it if necessary.
+
+Example:
 
 >
-    :lua vim.cmd('e'..vim.lsp.get_log_path())
+    configs.texlab.buf_build = buf_build
 <
+==============================================================================
+ADDING NEW SERVERS                                    *lspconfig-adding-servers*
+
+The three steps for adding and enabling a a new server configuration are:
 
+- load the lspconfig module (note, this is a stylistic choice)
+>
+   local lspconfig = require 'lspconfig'
+<
+- define the configuration
 
-Most of the time, the reason for failure is present in the logs.
+>
+   local configs = require 'lspconfig/configs'
+
+   -- Check if the config is already defined (useful when reloading this file)
+   if not lspconfig.foo_lsp then
+     configs.foo_lsp = {
+       default_config = {
+         cmd = {'/home/neovim/lua-language-server/run.sh'};
+         filetypes = {'lua'};
+         root_dir = function(fname)
+           return lspconfig.util.find_git_ancestor(fname)
+         end;
+         settings = {};
+       };
+     }
+   end
+
+- call `setup()` to enable the FileType autocmd
+>
+   lspconfig.foo_lsp.setup{}
+<
 ==============================================================================
-9. Built-in commands                             *lspconfig-built-in-commands*
+INDEX                                                          *lspconfig-index*
```

Comment:
Does this need a separate section? This is just repeating two sentences from the top of the file -- if needed, I'd just add a tag to those? (Maybe renamed to `lspconfig-serverlist`, as "index" already has a different meaning in the context of LSP.)

Reply:
Justin suggested it (and the name). The main point was to have a direct helptag link to the section. I think a small amount of duplication is fine.

Reply:
I understand, but can't you add the tag to the paragraph in the introduction? That strikes me as serving the same purpose without adding a second "source of truth".